lib/CL/devices/cpu_dbk/dbk_khr_jpeg_cpu.c

```diff
@@ -0,0 +1,217 @@
+
+void *
+pocl_cpu_init_dbk_khr_jpeg_encode (void const *attributes, int *status)
+{
+
+  jpeg_encode_state_t *state
+    = (jpeg_encode_state_t *)calloc (1, sizeof (jpeg_encode_state_t));
+  state->tj_handle = tj3Init (TJINIT_COMPRESS);
+  if (state->tj_handle == NULL)
+    {
+      const char *err_str = tj3GetErrorStr (state->tj_handle);
+      POCL_MSG_ERR ("create compress tj_handle failed: %s (fatal? %d)\n",
+                    err_str, tj3GetErrorCode (state->tj_handle));
+      *status = CL_OUT_OF_RESOURCES;
+      return NULL;
+    }
+
+  int jpeg_status;
+  cl_dbk_attributes_exp_jpeg_encode *jpeg_attr
+    = (cl_dbk_attributes_exp_jpeg_encode *)attributes;
+  jpeg_status = tj3Set (state->tj_handle, TJPARAM_QUALITY, jpeg_attr->quality);
+  if (jpeg_status != 0)
+    {
+      POCL_MSG_ERR ("could not set jpeg quality: %s (fatal? %d)\n",
+                    tj3GetErrorStr (state->tj_handle),
+                    tj3GetErrorCode (state->tj_handle));
+      *status = CL_OUT_OF_RESOURCES;
+      return NULL;
+    }
+
+  /* TODO: should this be configurable from the attributes? */
+  jpeg_status = tj3Set (state->tj_handle, TJPARAM_SUBSAMP, TJSAMP_420);
+  if (jpeg_status != 0)
+    {
+      POCL_MSG_ERR2 ("could not set jpeg subsampling", "%s",
+                     tj3GetErrorStr (state->tj_handle));
+      *status = CL_OUT_OF_RESOURCES;
+      return NULL;
+    }
+  /* todo set jpeg no_realloc */
+
+  return state;
+}
+
+int
+pocl_cpu_destroy_dbk_khr_jpeg_encode (void **kernel_data)
+{
+  jpeg_encode_state_t *state = (jpeg_encode_state_t *)*kernel_data;
+
+  tj3Destroy (state->tj_handle);
+  tj3Free (state->tj_buffer);
+  POCL_MEM_FREE (state);
+  return CL_SUCCESS;
+}
+
+int
+pocl_cpu_execute_dbk_khr_jpeg_decode (cl_program program,
+                                      cl_kernel kernel,
+                                      pocl_kernel_metadata_t *meta,
+                                      cl_uint dev_i,
+                                      struct pocl_argument *arguments)
+{
+  cl_device_id dev = program->devices[dev_i];
+  unsigned mem_id = dev->global_mem_id;
+  void *jpeg = pocl_cpu_get_ptr (&arguments[0], mem_id);
+  uint64_t *jpeg_size = pocl_cpu_get_ptr (&arguments[1], mem_id);
+  void *image = pocl_cpu_get_ptr (&arguments[2], mem_id);
+
+  /* make sure size is valid */
+  assert (*jpeg_size > 0);
+
+  jpeg_encode_state_t *state = kernel->data[dev_i];
+
+  int status;
+  status = tj3DecompressHeader (state->tj_handle, jpeg, *jpeg_size);
+  if (status != 0)
+    {
+      POCL_MSG_ERR ("tj3DecompressHeader failed: %s (fatal? %d)\n",
+                    tj3GetErrorStr (state->tj_handle),
+                    tj3GetErrorCode (state->tj_handle));
+      return CL_OUT_OF_RESOURCES;
+    }
+
+  int height = tj3Get (state->tj_handle, TJPARAM_JPEGHEIGHT);
+  int width = tj3Get (state->tj_handle, TJPARAM_JPEGWIDTH);
+  cl_mem jpeg_mem = *(cl_mem *)(arguments[2].value);
+  if (jpeg_mem->size < width * height * tjPixelSize[TJCS_RGB])
```

Comment:
```suggestion
  size_t height = tj3Get (state->tj_handle, TJPARAM_JPEGHEIGHT);
  size_t width = tj3Get (state->tj_handle, TJPARAM_JPEGWIDTH);
  cl_mem jpeg_mem = *(cl_mem *)(arguments[2].value);
  if (jpeg_mem->size < width * height * tjPixelSize[TJCS_RGB])
```
To avoid undefined behavior in `width * height * tjPixelSize` expression when the `width` and `height` variables are sufficiently large (e.g. `width == height == 65535`).

Reply:
it should be go fine since jpeg_mem->size is of size_t, but this does make it more clear what is happening

Reply:
The expression is evaluated as `int` before and then promoted to `size_t`for the comparison. See: https://godbolt.org/z/a9Wx13Pcz.